daphne_worker/src/dap.rs

```diff
@@ -251,11 +269,78 @@ where
         &self.global_config
     }
 
-    async fn get_task_config_for(
+    async fn get_task_config_considering_taskprov(
         &'srv self,
         task_id: Cow<'req, Id>,
+        metadata: Option<&ReportMetadata>,
     ) -> std::result::Result<Option<GuardedDapTaskConfig<'req>>, DapError> {
-        self.get_task_config(task_id).await.map_err(dap_err)
+        let found = self
+            .get_task_config(task_id.clone())
+            .await
+            .map_err(dap_err)?;
+        if found.is_some() {
+            return Ok(found);
+        }
+        // Not found and no error.
+        if metadata.is_none() {
+            // No report metadata, so we're not going to find anything.
+            return Ok(None);
+        }
+        let metadata_ref = metadata.unwrap();
+        let taskprov_task_config = get_taskprov_task_config(task_id.as_ref(), metadata_ref)?;
+        if taskprov_task_config.is_some() {
+            let global_config = self.get_global_config();
+            if !global_config.allow_taskprov {
+                // XXXRTH if DAP gets a generic denied error, we should use it here.
+                return Err(bad_request("taskprov is not allowed"));
+            }
+            if self.taskprov_hpke_collector_config.is_none() {
+                return Err(bad_request("No HPKE collector config for taskprov"));
+            }
+            if self.taskprov_vdaf_verify_key_init.is_none() {
+                return Err(bad_request("No VDAF verify key init config for taskprov"));
+            }
+            if self.taskprov_leader_bearer_token.is_none() {
+                return Err(bad_request("No leader bearer token for taskprov"));
+            }
+            if self.taskprov_collector_bearer_token.is_none() {
+                return Err(bad_request("No collector bearer token for taskprov"));
+            }
+
+            // We currently always opt-in; any additional checks that would lead to
+            // us opting out should be done here.
+
+            let taskprov_task_id = task_id.as_ref().clone();
+            let task_config = make_dap_task_config(
+                &taskprov_task_id,
+                taskprov_task_config.unwrap(),
+                self.taskprov_vdaf_verify_key_init.as_ref().unwrap(),
+                self.taskprov_hpke_collector_config.as_ref().unwrap(),
+            )?;
+
+            // Define a leader bearer token.  We do this so authorize_with_bearer_token()
+            // finds something.  XXXRTH Note that this is generating KV garbage that will
+            // need collection at some point.
+            self.set_leader_bearer_token(
```

Comment:
Why do we need to store the taskprov bearer tokens in KV? Couldn't we just grab it from the DaphneWorkerConfig as needed?